src/coreclr/src/vm/class.cpp

```diff
@@ -1184,107 +1184,112 @@ void ClassLoader::ValidateMethodsWithCovariantReturnTypes(MethodTable* pMT)
         return;
 
     // Step 1: Validate compatibility of return types on overriding methods
-
-    for (WORD i = 0; i < pParentMT->GetNumVirtuals(); i++)
+    if (pMT->GetClass()->HasCovariantOverride())
     {
-        MethodDesc* pMD = pMT->GetMethodDescForSlot(i);
-        MethodDesc* pParentMD = pParentMT->GetMethodDescForSlot(i);
+        for (WORD i = 0; i < pParentMT->GetNumVirtuals(); i++)
+        {
+            MethodDesc* pMD = pMT->GetMethodDescForSlot(i);
```

Comment:
You may be able to compare the slot pointers as a shortcut before resolving them back to MethodDesc.